src/cascadia/PublicTerminalCore/HwndTerminalAutomationPeer.cpp

```diff
@@ -0,0 +1,133 @@
+// Return Value:
+// - true, if the text is readable. false, otherwise.
+static constexpr bool IsReadable(std::wstring_view text)
+{
+    for (const auto c : text)
+    {
+        if (c > UNICODE_SPACE)
+        {
+            return true;
+        }
+    }
+    return false;
+}
+
+void HwndTerminalAutomationPeer::RecordKeyEvent(const WORD vkey)
+{
+    if (const auto charCode{ MapVirtualKey(vkey, MAPVK_VK_TO_CHAR) })
+    {
+        if (const auto keyEventChar{ gsl::narrow_cast<wchar_t>(charCode) }; IsReadable({ &keyEventChar, 1 }))
+        {
+            _keyEvents.emplace_back(keyEventChar);
+        }
+    }
+}
+
+// Method Description:
+// - Signals the ui automation client that the terminal's selection has changed and should be updated
+// Arguments:
+// - <none>
+// Return Value:
+// - <none>
+void HwndTerminalAutomationPeer::SignalSelectionChanged()
+{
+    UiaTracing::Signal::SelectionChanged();
+    LOG_IF_FAILED(UiaRaiseAutomationEvent(this, UIA_Text_TextSelectionChangedEventId));
+}
+
+// Method Description:
+// - Signals the ui automation client that the terminal's output has changed and should be updated
+// Arguments:
+// - <none>
+// Return Value:
+// - <none>
+void HwndTerminalAutomationPeer::SignalTextChanged()
+{
+    UiaTracing::Signal::TextChanged();
+    LOG_IF_FAILED(UiaRaiseAutomationEvent(this, UIA_Text_TextChangedEventId));
+}
+
+// Method Description:
+// - Signals the ui automation client that the cursor's state has changed and should be updated
+// Arguments:
+// - <none>
+// Return Value:
+// - <none>
+void HwndTerminalAutomationPeer::SignalCursorChanged()
+{
+    UiaTracing::Signal::CursorChanged();
+    LOG_IF_FAILED(UiaRaiseAutomationEvent(this, UIA_Text_TextSelectionChangedEventId));
+}
+
+void HwndTerminalAutomationPeer::NotifyNewOutput(std::wstring_view newOutput)
+{
+    // Try to suppress any events (or event data)
+    // that is just the keypress the user made
+    auto sanitized{ Sanitize(newOutput) };
+    while (!_keyEvents.empty() && IsReadable(sanitized))
+    {
+        if (til::toupper_ascii(sanitized.front()) == _keyEvents.front())
+        {
+            // the key event's character (i.e. the "A" key) matches
+            // the output character (i.e. "a" or "A" text).
+            // We can assume that the output character resulted from
+            // the pressed key, so we can ignore it.
+            sanitized = sanitized.substr(1);
+            _keyEvents.pop_front();
+        }
+        else
+        {
+            // The output doesn't match,
+            // so clear the input stack and
+            // move on to fire the event.
+            _keyEvents.clear();
+            break;
+        }
+    }
+
+    // Suppress event if the remaining text is not readable
+    if (!IsReadable(sanitized))
+    {
+        return;
+    }
+
+    const auto sanitizedBstr = SysAllocString(sanitized.c_str());
+    static auto activityId = SysAllocString(L"TerminalTextOutput");

```

Comment:
This is all copied over too, so I'll probably add a helper function that does just this.

Reply:
Just duplicate it for now. :)